### ISSUE TITLE
Feature: Possibility for rename resources, snapshots and storage pools

### ISSUE THREAD

**Issue by kvaps** [LINBIT/linstor-server#44]
According #17 I've used two different storage pools and call auto-placing command two times, for place data on two different racks.

But few time ago there was released linstor CSI plugin which allows to set AUX properties: `ReplicasOnSame` and `ReplicasOnDifferent`, what made my setup not so optimal as just using AUX properties.

I need to rename my current storage pools for allow using same pool name but with differnet aux properties.

**Comment by ghernadi** [comment]
Currently Linstor does not support any kind of renaming action. 

If you really have to do this, you have two options: 
* NOT RECOMMENDED: you do it manually by editing the database (NOT recommended, as you have to be very careful to update all references to the changed name) 
* a bit "more" recommended: delete your current database (or perform "lostNode" commands for the affected nodes) and recreate using migration everything with the new storage pool names. 

Note: we do have internal discussions if and how we could provide renaming actions. That means this situation might be easier to resolve in (far) future.

**Comment by kvaps** [comment]
Just mention that resource-groups also needs this feature

**Comment by volkerjaenisch** [comment]
@raltnoeder 
So this looks like a piece of not well designed software to me. Renaming of resources is not a brand new feature, but an really old issue. In DB terms no one would use the name of a resource to reference it. You would use a kind of primary key to share the resource it with other entities. So that you can easily rename the resource.
In my case I do not use LVM and ZFS. So in the end it comes down to linstore and drbd which are both from the same company.
And at least these two bits of software could be improved for renaming resources.
Cheers,
Volker

**Comment by raltnoeder** [comment]
I designed the software (LINSTOR), and the main priority when the project was started in 2016 was better reliability and maintainability, because the typical use case for DRBD is for higher availability. Availability is a function of the time it takes to resolve an outage, or to perform maintenance on a system. Complexity is the natural enemy of administrators as well as software developers, and that includes data that is full of surrogate keys consisting of meaningless numbers and letters.
DRBD did not have the ability to rename resources back then, and this functionality was only added very recently, and even now is not without corner-cases and limitations (and also limited to the local node, it can't rename across the entire cluster in a transaction). For example, resources are associated with various paths in the Linux filesystem, e.g. in /dev, but also in /sys, and those paths can not be changed while any process is accessing it.
If you think that's not well-designed, then yes, maybe it isn't, but the reason for that is not the way data is stored in the database (that would just be a simple update statement either way), but the fact that the entire environment around LINSTOR isn't well-designed. We are trying to build a cluster-wide transaction-safe storage automation system on top of an operating system and user environment that does not even provide node-local transaction-safety and is full of corner-cases, potential failure modes, undocumented failure modes, race conditions and similar problems. Linstor contains tens of thousands of lines of code just to deal with reacting to and correcting those problems. And this is the main reason why we can't just add a feature that makes a cluster-wide change in a day and 50 lines of code.

**Comment by raltnoeder** [comment]
The strong coupling between resource names in LINSTOR, DRBD and the backing volumes exists for maintainability. Most administrators are already overwhelmed with using the system as it is, and it seems that most people don't read the manual. Now imagine an administrator trying to fix 900 resources on a system where he or she has to deal with at least 3 different names for the different layers of the storage system - LINSTOR, DRBD, backing storage - trying to figure out which DRBD resource belongs to which LINSTOR resource and why it doesn't attach to the backing storage that has yet another name. Whenever that happens, it's simply a downtime multiplier.
There also did not seem to be a reason to rename resources as a part of daily business - after all, DRBD did not have the ability to rename resources at runtime for more than 15 years, and apparently, noone was missing that feature. In the few cases where a major change to infrastructure would suggest renaming a resource to reflect the change, a manual intervention would have seemed sufficient. If lots of resources need to be renamed all the time as a part of daily business, that's quite certainly the result of bad planning.

**Comment by luissimas** [comment]
Hello folks! I know this is an old issue and some valid points were made in the discussions here as to why this is far from trivial to implement. Still, I'd like to know if there are any plans to implement the ability to rename some of LINSTOR's objects, particularly user-facing ones. 

We're implementing a LINSTOR driver for Incus (https://github.com/lxc/incus/pull/1621), and we have the need for renaming volumes and snapshots (in Incus's terminology). To be able to implement that, it would be really nice to have the ability to rename resource definitions and snapshots in LINSTOR. 

Without the renaming capabilities in LINSTOR, we end up having to associate IDs with resource definitions and snapshots, which end up increasing the complexity of the overall solution and also makes it harder for operators to correlate Incus volume names with LINSTOR's resource definition names.

Is there any chance to implement the renaming capability for those higher-level objects?

**Comment by ghernadi** [comment]
No, there are no plans in this regard. However, you might want to try setting aux-properties on a resource-definition stating the name within Inucs. For example `linstor rd sp $linstor_rsc_dfn_name --aux InucsName $incus_rsc_name`. If the plugin fetches the list of resource-definitions it also always gets the properties. Which means that the plugin would need to check `rd.props["Aux/IncusName"]` instead of `rd.name` but I think this might be a compromise. For humans you can also use `linstor rd l -s Aux/IncusName`  to tell the LINSTOR client to show an additional column with the values of `Aux/IncusName` in the table of resource-definitions.

Right now you cannot set any kind of properties on LINSTOR snapshots, but that is something we can talk about to add to LINSTOR as well as the `--show-props` option when listing snapshots and storage pools (to kinda address the first comment also).

Note: If you are exploring LINSTOR's API, please do not try to use the "external name" of resource-definition for 2 reasons: 1) You cannot change those either and 2) We are thinking about deprecating and later remove that anyways.

**Comment by luissimas** [comment]
Thanks for the suggestions!

Yeah, we're currently relying on aux-properties to set the name and some extra metadata on the resource definitions. However, in our use case, we'd like to fetch a single resource definition from LINSTOR based on its name on Incus, and I don't think we can efficiently query LINSTOR based on an aux-property.

Setting properties on LINSTOR snapshots would be a great addition, as we have the exact same use cases that would greatly benefit from the ability to store metadata for snapshots in aux-properties.

**Comment by rp-** [comment]
> Yeah, we're currently relying on aux-properties to set the name and some extra metadata on the resource definitions. However, in our use case, we'd like to fetch a single resource definition from LINSTOR based on its name on Incus, and I don't think we can efficiently query LINSTOR based on an aux-property.

You can do that:

```
root@cs-121-agent:~# linstor --curl rd l --props Aux/cs-vm-name=r-4-VM
curl -X GET http://localhost:3370/v1/resource-definitions?props=Aux/cs-vm-name=r-4-VM
```

**Comment by rp-** [comment]
As this would be a controller only operation, even with 10 000 resource-definitions I would assume < 100ms.
Linstor reads all data into RAM, so there is no disk database lookup.

**Comment by luissimas** [comment]
Ok, that's good to know. Thanks a lot, we'll consider that approach in our implementation.
